trin-execution/src/walkers/trie_walker.rs

```diff
@@ -0,0 +1,378 @@
+            path: vec![],
+            proof: vec![root_value],
+        };
+
+        Ok(Self {
+            is_partial_trie: false,
+            trie,
+            stack: vec![root_proof],
+        })
+    }
+
+    pub fn new_partial_trie(root_hash: B256, trie: TrieDB) -> anyhow::Result<Self> {
+        // if the storage root is empty then there is no storage to gossip
+        if root_hash == EMPTY_ROOT_HASH {
+            return Ok(Self {
+                is_partial_trie: true,
+                trie: Arc::new(trie),
+                stack: vec![],
+            });
+        }
+
+        let root_value = match trie
+            .get(root_hash.as_slice())
+            .expect("Failed to read node from")
+        {
+            Some(root_value) => root_value.into(),
+            None => {
+                // The trie db is empty so we can't walk it return an empty iterator
+                return Ok(Self {
+                    is_partial_trie: true,
+                    trie: Arc::new(trie),
+                    stack: vec![],
+                });
+            }
+        };
+
+        let root_proof = TrieProof {
+            path: vec![],
+            proof: vec![root_value],
+        };
+
+        Ok(Self {
+            is_partial_trie: true,
+            trie: Arc::new(trie),
+            stack: vec![root_proof],
+        })
+    }
+
+    fn process_node(
+        &mut self,
+        node: Node,
+        partial_proof: Vec<Bytes>,
+        path: Vec<u8>,
+    ) -> anyhow::Result<()> {
+        // We only need to process hash nodes, because if the node isn't a hash node the leaf is
+        // already embedded in the proof
+        if let Node::Hash(hash) = node {
+            let value_result = self
+                .trie
+                .get(hash.hash.as_slice())
+                .expect("Failed to read node from the database");
+
+            let value = match value_result {
+                Some(value) => value,
+                None => {
+                    // If we are walking a partial trie, some nodes won't be available in the
+                    // database
+                    if self.is_partial_trie {
+                        return Ok(());
+                    }
+                    return Err(anyhow::anyhow!("Node not found in the database"));
+                }
+            };
+            let decoded_node = decode_node(&mut value.as_ref())?;
+            match decoded_node {
+                Node::Leaf(_) | Node::Extension(_) | Node::Branch(_) => {
+                    self.stack.push(TrieProof {
+                        path,
+                        proof: [partial_proof, vec![value.into()]].concat(),
+                    });
+                }
+                // can't be a hash node because we just decoded it
+                Node::Empty | Node::Hash(_) => (),
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<TrieDB: DB> Iterator for TrieWalker<TrieDB> {
+    type Item = TrieProof;
```

Comment:
I added a commit for why it would panic, the only reason it would panic is if the trie database was corrupted or a database read failed, which at that point you wouldn't be able to continue anyways as both of those are unrecoverable states.